src/parser.rs

```diff
@@ -407,6 +411,243 @@ pub fn parse_gcov(gcov_path: &Path) -> Result<Vec<(String,CovResult)>, ParserErr
     Ok(results)
 }
 
+fn get_xml_attribute(attributes: &Vec<OwnedAttribute>, name: &str) -> Option<String> {
+    for a in attributes {
+        if a.name.local_name.as_str() == name {
+            return Some(String::from(a.value.clone()));
+        }
+    }
+    None
+}
+
+fn parse_java_sourcefile<T: Read>(parser: &mut EventReader<T>) -> Result<(BTreeMap<u32,u64>, BTreeMap<(u32,u32),bool>), ParserError> {
+    let mut lines: BTreeMap<u32,u64> = BTreeMap::new();
+    let mut branches: BTreeMap<(u32,u32),bool> = BTreeMap::new();
+
+    loop {
+        match parser.next() {
+            Ok(XmlEvent::StartElement { name, attributes, .. }) => {
+                match name.local_name.as_str() {
+                    "line" => {
+                        // Save info about source file line
+                        let ci = get_xml_attribute(&attributes, "ci").unwrap().parse::<u64>().unwrap();
+                        let cb = get_xml_attribute(&attributes, "cb").unwrap().parse::<u64>().unwrap();
+                        let mb = get_xml_attribute(&attributes, "mb").unwrap().parse::<u64>().unwrap();
+                        let nr = get_xml_attribute(&attributes, "nr").unwrap().parse::<u32>().unwrap();
+
+                        if mb > 0 || cb > 0 {
+                            // Line is a branch
+                            for branch in 0..cb {
+                                branches.insert((nr, branch as u32), true);
+                            }
+
+                            for branch in cb..cb+mb{
+                                branches.insert((nr, branch as u32), false);
+                            }
+                        } else {
+                            // Line is a statement
+                            lines.insert(nr, ci);
```

Comment:
Is `ci` the number of times the instruction was executed?

Reply:
No, it's the count of covered instructions on that given code line.

Reply:
OK. In Result, the lines map is `line number` -> `hit count`. We should have the same here.
Can you check with an example what happens in the XML file if you have a for loop that is executed multiple times?

Reply:
[The JaCoCo docs](https://www.jacoco.org/jacoco/trunk/doc/counters.html) state:

> Due to the fact that a single line typically compiles to multiple byte code instructions the source code highlighting shows three different status for each line containing source code:

>  -  No coverage: No instruction in the line has been executed (red background)
>  -  Partial coverage: Only a part of the instruction in the line have been executed (yellow background)
>  -  Full coverage: All instructions in the line have been executed (green background)

> Depending on source formatting a single line of a source code may refer to multiple methods or multiple classes. 

It's not the count, but covered instructions per line.

Reply:
For reference, https://github.com/jacoco/jacoco/issues/89.